ssl/quic/json_enc.c

```diff
@@ -0,0 +1,721 @@
+}
+
+static void json_indent(OSSL_JSON_ENC *json)
+{
+    size_t i, depth;
+
+    json->defer_indent = 0;
+
+    if (!in_pretty(json))
+        return;
+
+    json_write_char(json, '\n');
+
+    depth = json->stack_end_byte * 8 + json->stack_end_bit;
+    for (i = 0; i < depth * 4; ++i)
+        json_write_str(json, "    ");
+}
+
+static int json_pre_item(OSSL_JSON_ENC *json)
+{
+    int s;
+
+    if (ossl_json_in_error(json))
+        return 0;
+
+    switch (json->state) {
+    case STATE_PRE_COMMA:
+        s = json_peek(json);
+
+        if (s == 0) {
+            json_raise_error(json);
+            return 0;
+        }
+
+        if (s == 1) {
+            json_write_char(json, ',');
+            if (ossl_json_in_error(json))
+                return 0;
+
+            json_indent(json);
+        }
+
+        if (s < 0 && in_seq(json))
+            json_write_char(json, '\x1E');
+
+        json->state = STATE_PRE_ITEM;
+        break;
+
+    case STATE_PRE_ITEM:
+        break;
+
+    case STATE_PRE_KEY:
+    default:
+        json_raise_error(json);
+        return 0;
+    }
+
+    return 1;
+}
+
+static void json_post_item(OSSL_JSON_ENC *json)
+{
+    int s = json_peek(json);
+
+    json->state = STATE_PRE_COMMA;
+
+    if (s < 0 && in_seq(json))
+        json_write_char(json, '\n');
+}
+
+/*
+ * Begin a composite structure (object or array).
+ *
+ * type: 0=object, 1=array.
+ */
+static void composite_begin(OSSL_JSON_ENC *json, int type, char ch)
+{
+    if (!json_pre_item(json)
+        || !json_push(json, type))
+        json_raise_error(json);
+
+    json_write_char(json, ch);
+    json->defer_indent = 1;
+}
+
+/*
+ * End a composite structure (object or array).
+ *
+ * type: 0=object, 1=array. Errors on mismatch.
+ */
+static void composite_end(OSSL_JSON_ENC *json, int type, char ch)
+{
+    int was_defer = json->defer_indent;
+
+    if (ossl_json_in_error(json))
+        return;
+
+    json->defer_indent = 0;
+
+    if (json_peek(json) != type) {
+        json_raise_error(json);
+        return;
+    }
+
+    if (type == 0 && json->state == STATE_PRE_ITEM) {
+        json_raise_error(json);
+        return;
+    }
+
+    if (!json_pop(json)) {
+        json_raise_error(json);
+        return;
+    }
+
+    if (!was_defer)
+        json_indent(json);
+
+    json_write_char(json, ch);
+    json_post_item(json);
+}
+
+/* Begin a new JSON object. */
+void ossl_json_object_begin(OSSL_JSON_ENC *json)
+{
+    composite_begin(json, 0, '{');
+    json->state = STATE_PRE_KEY;
+}
+
+/* End a JSON obejct. Must be matched with a call to ossl_json_object_begin(). */
+void ossl_json_object_end(OSSL_JSON_ENC *json)
+{
+    composite_end(json, 0, '}');
+}
+
+/* Begin a new JSON array. */
+void ossl_json_array_begin(OSSL_JSON_ENC *json)
+{
+    composite_begin(json, 1, '[');
+    json->state = STATE_PRE_ITEM;
+}
+
+/* End a JSON array. Must be matched with a call to ossl_json_array_end(). */
+void ossl_json_array_end(OSSL_JSON_ENC *json)
```

Comment:
ossl_json_array_end must be matched with a call to ossl_json_array_end - that comment belongs against ossL_json_array_begin.